File: src/main/java/org/anhonesteffort/btc/strategy/impl/SimpleAskIdentifyingStrategy.java
/*
 * Copyright (C) 2016 An Honest Effort LLC.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.anhonesteffort.btc.strategy.impl;

import org.anhonesteffort.btc.book.Order;
import org.anhonesteffort.btc.http.request.RequestFactory;
import org.anhonesteffort.btc.state.GdaxState;
import org.anhonesteffort.btc.strategy.AskIdentifyingStrategy;
import org.anhonesteffort.btc.util.LongCaster;

import java.util.Optional;

public class SimpleAskIdentifyingStrategy extends AskIdentifyingStrategy {

  public SimpleAskIdentifyingStrategy(LongCaster caster, RequestFactory requests) {
    super(caster, requests);
  }

  @Override
  protected Optional<Double> identifyPrice(
      Order bidPosition, Optional<Order> lastAsk, GdaxState state, long nanoseconds
  ) {
    double bidFloor   = caster.toDouble(state.getOrderBook().getBidLimits().peek().get().getPrice());
    double askCeiling = caster.toDouble(state.getOrderBook().getAskLimits().peek().get().getPrice());
    double lastPrice  = lastAsk.isPresent() ? caster.toDouble(lastAsk.get().getPrice()) : -1l;
    double nextPrice  = (lastPrice - 0.01d);
    double bidPrice   = caster.toDouble(bidPosition.getPrice());

    if (!lastAsk.isPresent()) {
      return Optional.of(askCeiling);
    } else if (nextPrice > bidFloor && nextPrice > bidPrice) {
      return Optional.of(nextPrice);
    } else if (bidPrice > bidFloor) {
      return Optional.of(bidPrice);
    } else {
      return Optional.of(bidFloor + 0.01d);
    }
  }

}

File: jvm/src/main/scala/org/anhonesteffort/trading/strategy/impl/SimpleAskIdentifyingStrategy.scala
package org.anhonesteffort.trading.strategy.impl

import org.anhonesteffort.trading.book.Orders.Order
import org.anhonesteffort.trading.http.request.RequestFactory
import org.anhonesteffort.trading.state.GdaxState
import org.anhonesteffort.trading.strategy.AskIdentifyingStrategy

class SimpleAskIdentifyingStrategy(requests: RequestFactory) extends AskIdentifyingStrategy(requests) {

  override def identifyPrice(bidPosition: Order, lastAsk: Option[Order], state: GdaxState, ns: Long): Option[Double] = {
    val bidFloor   : Double = state.orderBook.getBidLimits.peek.get.getPrice
    val askCeiling : Double = state.orderBook.getAskLimits.peek.get.getPrice
    val lastPrice  : Double = lastAsk match { case None => -1d case Some(price) => price.getPrice}
    val nextPrice  : Double = lastPrice - 0.01d
    val bidPrice   : Double = bidPosition.getPrice

    if (lastAsk.isEmpty) {
      Some(askCeiling)
    } else if (nextPrice > bidFloor && nextPrice > bidPrice) {
      Some(nextPrice)
    } else if (bidPrice > bidFloor) {
      Some(bidPrice)
    } else {
      Some(bidFloor + 0.01d)
    }
  }

}
